// @paycan/sdk/dist/index.esm.js
/**
 * HTTP Client with automatic token management
 */
class HttpClient {
    constructor(config) {
        this.token = null;
        this.tokenExpiry = null;
        this.refreshPromise = null;
        this.config = config;
    }
    /**
     * Set the authentication token
     */
    setToken(token) {
        this.token = token;
        this.tokenExpiry = this.parseTokenExpiry(token);
        this.log('Token set, expires at:', new Date(this.tokenExpiry || 0).toISOString());
    }
    /**
     * Get current token
     */
    getToken() {
        return this.token;
    }
    /**
     * Check if token needs refresh
     */
    needsRefresh() {
        if (!this.tokenExpiry)
            return false;
        const threshold = (this.config.refreshThreshold || 300) * 1000; // Convert to ms
        const now = Date.now();
        const timeUntilExpiry = this.tokenExpiry - now;
        return timeUntilExpiry <= threshold;
    }
    /**
     * Parse JWT token to get expiry time
     */
    parseTokenExpiry(token) {
        try {
            const payload = token.split('.')[1];
            const decoded = JSON.parse(atob(payload));
            return decoded.exp ? decoded.exp * 1000 : null; // Convert to ms
        }
        catch (error) {
            this.log('Failed to parse token expiry:', error);
            return null;
        }
    }
    /**
     * Make HTTP request with automatic token refresh
     */
    async request(endpoint, options = {}) {
        await this.ensureValidToken();
        const url = `${this.config.apiUrl}${endpoint}`;
        const headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
            ...(options.headers || {}),
        };
        if (this.token) {
            headers['Authorization'] = `Bearer ${this.token}`;
        }
        this.log(`${options.method || 'GET'} ${url}`);
        const response = await fetch(url, {
            ...options,
            headers,
        });
        return this.handleResponse(response);
    }
    /**
     * Ensure we have a valid token, refresh if needed
     */
    async ensureValidToken() {
        if (!this.token) {
            throw new Error('Not authenticated. Please call setUser() first.');
        }
        if (this.needsRefresh()) {
            // If refresh is already in progress, wait for it
            if (this.refreshPromise) {
                await this.refreshPromise;
                return;
            }
            // Start refresh
            this.refreshPromise = this.refreshToken();
            await this.refreshPromise;
            this.refreshPromise = null;
        }
    }
    /**
     * Refresh the authentication token
     * This is called automatically when token is close to expiry
     */
    async refreshToken() {
        this.log('Refreshing token...');
        // Token refresh happens through the auth endpoint
        // The backend should handle token refresh logic
        // For now, we'll rely on the token's expiry time
        // If you implement a refresh endpoint, call it here
    }
    /**
     * Handle HTTP response
     */
    async handleResponse(response) {
        const contentType = response.headers.get('content-type');
        const isJson = contentType?.includes('application/json');
        if (!response.ok) {
            let error = {
                message: response.statusText,
                status: response.status,
            };
            if (isJson) {
                try {
                    const errorData = await response.json();
                    error = {
                        message: errorData.message || error.message,
                        errors: errorData.errors,
                        status: response.status,
                    };
                }
                catch (e) {
                    // Use default error if JSON parsing fails
                }
            }
            this.log('API Error:', error);
            throw error;
        }
        if (isJson) {
            return response.json();
        }
        return {};
    }
    /**
     * Debug logging
     */
    log(...args) {
        if (this.config.debug) {
            console.log('[PayCan SDK]', ...args);
        }
    }
    /**
     * GET request
     */
    async get(endpoint, params) {
        let url = endpoint;
        if (params) {
            const queryString = new URLSearchParams(Object.entries(params).reduce((acc, [key, value]) => {
                if (value !== undefined && value !== null) {
                    acc[key] = String(value);
                }
                return acc;
            }, {})).toString();
            if (queryString) {
                url = `${endpoint}?${queryString}`;
            }
        }
        return this.request(url, { method: 'GET' });
    }
    /**
     * POST request
     */
    async post(endpoint, data) {
        return this.request(endpoint, {
            method: 'POST',
            body: data ? JSON.stringify(data) : undefined,
        });
    }
    /**
     * PUT request
     */
    async put(endpoint, data) {
        return this.request(endpoint, {
            method: 'PUT',
            body: data ? JSON.stringify(data) : undefined,
        });
    }
    /**
     * DELETE request
     */
    async delete(endpoint) {
        return this.request(endpoint, { method: 'DELETE' });
    }
}

/**
 * Orders Resource
 *
 * Handle order-related operations
 */
class Orders {
    constructor(http, config) {
        this.http = http;
        this.ordersCache = null;
        this.cacheTimestamp = null;
        this.CACHE_KEY = 'paycan_orders_cache';
        this.CACHE_TS_KEY = 'paycan_orders_cache_timestamp';
        const cacheTTLSeconds = config.cacheTtl ?? 60;
        this.cacheTtl = cacheTTLSeconds * 1000; // Convert to milliseconds
        // Load cache from localStorage on initialization
        this.loadCacheFromStorage();
    }
    /**
     * Load cache from localStorage
     */
    loadCacheFromStorage() {
        if (typeof window === 'undefined' || typeof localStorage === 'undefined') {
            return;
        }
        try {
            const cached = localStorage.getItem(this.CACHE_KEY);
            const timestamp = localStorage.getItem(this.CACHE_TS_KEY);
            if (cached && timestamp) {
                const ts = parseInt(timestamp, 10);
                if (Date.now() - ts < this.cacheTtl) {
                    this.ordersCache = JSON.parse(cached);
                    this.cacheTimestamp = ts;
                }
                else {
                    this.clearCache();
                }
            }
        }
        catch (error) {
            this.clearCache();
        }
    }
    /**
     * Save cache to localStorage
     */
    saveCacheToStorage() {
        if (typeof window === 'undefined' || typeof localStorage === 'undefined') {
            return;
        }
        try {
            if (this.ordersCache && this.cacheTimestamp) {
                localStorage.setItem(this.CACHE_KEY, JSON.stringify(this.ordersCache));
                localStorage.setItem(this.CACHE_TS_KEY, this.cacheTimestamp.toString());
            }
        }
        catch (error) {
            // Silently fail if localStorage quota exceeded or unavailable
        }
    }
    /**
     * Clear the orders cache
     *
     * Note: Cache is automatically cleared when creating new orders or completing checkouts
     * to ensure fresh data is fetched on next request.
     */
    clearCache() {
        this.ordersCache = null;
        this.cacheTimestamp = null;
        if (typeof window !== 'undefined' && typeof localStorage !== 'undefined') {
            try {
                localStorage.removeItem(this.CACHE_KEY);
                localStorage.removeItem(this.CACHE_TS_KEY);
            }
            catch (error) {
                // Silently fail
            }
        }
    }
    /**
     * List all orders for the authenticated user
     *
     * @param params - Query parameters for filtering, sorting, pagination, and includes
     * @param params.filter - Filter orders (status, gateway, order_number, created_after, created_before)
     * @param params.include - Include related data (product, productPrice, productPrice.product, transactions, fulfillments, subscription)
     * @param params.sort - Sort field (created_at, total, status) - prefix with - for descending
     * @param params.per_page - Items per page (1-100, default: 15)
     * @param params.page - Page number
     *
     * @example
     * // Basic usage
     * const orders = await paycan.orders.list();
     *
     * @example
     * // With filtering and sorting
     * const completedOrders = await paycan.orders.list({
     *   filter: { status: 'completed', created_after: '2024-01-01' },
     *   include: 'productPrice.product',
     *   sort: '-created_at',
     *   per_page: 20
     * });
     */
    async list(params) {
        const result = await this.http.get('/api/user/orders', params);
        this.ordersCache = result.data;
        this.cacheTimestamp = Date.now();
        this.saveCacheToStorage();
        return result;
    }
    /**
     * Get a specific order by ID
     *
     * @param orderId - The order ID
     * @param params - Query parameters
     * @param params.include - Include related data (product, productPrice, productPrice.product, transactions, fulfillments, subscription)
     *
     * @example
     * const order = await paycan.orders.get('order-id-123');
     *
     * @example
     * // With includes
     * const order = await paycan.orders.get('order-id-123', {
     *   include: 'productPrice.product,transactions,fulfillments'
     * });
     */
    async get(orderId, params) {
        return this.http.get(`/api/user/orders/${orderId}`, params);
    }
    /**
     * Get download links for an order
     *
     * @example
     * const downloads = await paycan.orders.getDownloads('order-id-123');
     */
    async getDownloads(orderId) {
        return this.http.get(`/api/user/orders/${orderId}/downloads`);
    }
    /**
     * Get license keys for an order
     *
     * @example
     * const licenses = await paycan.orders.getLicenses('order-id-123');
     */
    async getLicenses(orderId) {
        return this.http.get(`/api/user/orders/${orderId}/licenses`);
    }
}

/**
 * Subscriptions Resource
 *
 * Handle subscription-related operations
 */
class Subscriptions {
    constructor(http, config) {
        this.http = http;
        this.subscriptionsCache = null;
        this.cacheTimestamp = null;
        this.CACHE_KEY = 'paycan_subscriptions_cache';
        this.CACHE_TS_KEY = 'paycan_subscriptions_cache_timestamp';
        // Get cache TTL from config (default 60 seconds), convert to milliseconds
        const cacheTTLSeconds = config.cacheTtl ?? 60;
        this.cacheTtl = cacheTTLSeconds * 1000;
        // Load cache from localStorage on initialization
        this.loadCacheFromStorage();
    }
    /**
     * Load cache from localStorage
     */
    loadCacheFromStorage() {
        if (typeof window === 'undefined' || typeof localStorage === 'undefined') {
            return;
        }
        try {
            const cached = localStorage.getItem(this.CACHE_KEY);
            const timestamp = localStorage.getItem(this.CACHE_TS_KEY);
            if (cached && timestamp) {
                this.subscriptionsCache = JSON.parse(cached);
                this.cacheTimestamp = parseInt(timestamp, 10);
            }
        }
        catch (error) {
            // Silently fail if localStorage is unavailable or data is corrupted
            this.clearCache();
        }
    }
    /**
     * Save cache to localStorage
     */
    saveCacheToStorage() {
        if (typeof window === 'undefined' || typeof localStorage === 'undefined') {
            return;
        }
        try {
            if (this.subscriptionsCache && this.cacheTimestamp) {
                localStorage.setItem(this.CACHE_KEY, JSON.stringify(this.subscriptionsCache));
                localStorage.setItem(this.CACHE_TS_KEY, this.cacheTimestamp.toString());
            }
        }
        catch (error) {
            // Silently fail if localStorage quota exceeded or unavailable
        }
    }
    /**
     * Clear the subscriptions cache
     * Call this after subscription changes (cancel, resume, upgrade)
     */
    clearCache() {
        this.subscriptionsCache = null;
        this.cacheTimestamp = null;
        if (typeof window !== 'undefined' && typeof localStorage !== 'undefined') {
            try {
                localStorage.removeItem(this.CACHE_KEY);
                localStorage.removeItem(this.CACHE_TS_KEY);
            }
            catch (error) {
                // Silently fail
            }
        }
    }
    /**
     * Get cached subscriptions or fetch fresh data
     * PERFORMANCE TIP: This method caches subscriptions to avoid rate limits
     * and improve performance when checking multiple products/prices.
     * Cache can be disabled by setting cacheTtl to 0.
     */
    async getCachedSubscriptions() {
        // If cache is disabled (TTL = 0), always fetch fresh data
        if (this.cacheTtl === 0) {
            const response = await this.list();
            return response.data || [];
        }
        const now = Date.now();
        const isCacheValid = this.subscriptionsCache &&
            this.cacheTimestamp &&
            (now - this.cacheTimestamp) < this.cacheTtl;
        if (isCacheValid) {
            return this.subscriptionsCache;
        }
        // Fetch fresh data
        const response = await this.list();
        this.subscriptionsCache = response.data || [];
        this.cacheTimestamp = now;
        this.saveCacheToStorage();
        return this.subscriptionsCache;
    }
    /**
     * List all subscriptions for the authenticated user
     *
     * @param params - Query parameters for filtering, sorting, pagination, and includes
     * @param params.filter - Filter subscriptions (status, gateway, created_after, created_before)
     * @param params.include - Include related data (product, productPrice, productPrice.product, order, transactions)
     * @param params.sort - Sort field (created_at, status, next_billing_date) - prefix with - for descending
     * @param params.per_page - Items per page (1-100, default: 15)
     * @param params.page - Page number
     *
     * @example
     * // Basic usage
     * const subscriptions = await paycan.subscriptions.list();
     *
     * @example
     * // With filtering and sorting
     * const activeSubscriptions = await paycan.subscriptions.list({
     *   filter: { status: 'active', gateway: 'stripe' },
     *   include: 'productPrice.product',
     *   sort: '-created_at',
     *   per_page: 20
     * });
     */
    async list(params) {
        // Clear cache when explicitly listing to ensure fresh data
        this.clearCache();
        return this.http.get('/api/user/subscriptions', params);
    }
    /**
     * List all active subscriptions (uses cache)
     *
     * @example
     * const activeSubscriptions = await paycan.subscriptions.listActive();
     */
    async listActive() {
        const subscriptions = await this.getCachedSubscriptions();
        return subscriptions.filter((sub) => sub.status === 'active');
    }
    /**
     * Get a specific subscription by ID
     * Checks cache first before making API request
     *
     * @param subscriptionId - The subscription ID
     * @param params - Query parameters
     * @param params.include - Include related data (product, productPrice, productPrice.product, order, transactions)
     *
     * @example
     * const subscription = await paycan.subscriptions.get('sub-123');
     *
     * @example
     * // With includes
     * const subscription = await paycan.subscriptions.get('sub-123', {
     *   include: 'productPrice.product,order,transactions'
     * });
     */
    async get(subscriptionId, params) {
        // Check cache first (only if no specific includes requested)
        if (!params?.include) {
            const cachedSubscriptions = await this.getCachedSubscriptions();
            const cached = cachedSubscriptions.find((sub) => sub.id === subscriptionId);
            if (cached) {
                return { data: cached };
            }
        }
        // Not in cache or specific includes requested, fetch from API
        return this.http.get(`/api/user/subscriptions/${subscriptionId}`, params);
    }
    /**
     * Cancel a subscription
     *
     * @example
     * await paycan.subscriptions.cancel('sub-123');
     */
    async cancel(subscriptionId) {
        const result = await this.http.post(`/api/user/subscriptions/${subscriptionId}/cancel`);
        // Clear cache after mutation
        this.clearCache();
        return result;
    }
    /**
     * Resume a cancelled subscription
     *
     * @example
     * await paycan.subscriptions.resume('sub-123');
     */
    async resume(subscriptionId) {
        const result = await this.http.post(`/api/user/subscriptions/${subscriptionId}/resume`);
        // Clear cache after mutation
        this.clearCache();
        return result;
    }
    /**
     * Change subscription plan
     *
     * @example
     * await paycan.subscriptions.change('sub-123', { new_price_id: 'price-456' });
     */
    async change(subscriptionId, data) {
        const result = await this.http.post(`/api/user/subscriptions/${subscriptionId}/change`, data);
        // Clear cache after mutation
        this.clearCache();
        return result;
    }
}

/**
 * Checkout Resource
 *
 * Handle checkout and payment portal operations
 */
class Checkout {
    constructor(http, orders, subscriptions) {
        this.http = http;
        this.orders = orders;
        this.subscriptions = subscriptions;
    }
    /**
     * Create a checkout session
     *
     * @example
     * const session = await paycan.checkout.create({
     *   product_id: 1,
     *   product_price_id: 1,
     *   gateway: 'stripe',
     *   billing_email: 'user@example.com', // Optional for authenticated users
     *   billing_name: 'John Doe',          // Optional for authenticated users
     *   quantity: 1                         // Optional, defaults to 1
     * });
     * window.location.href = session.checkout_url;
     */
    async create(data) {
        const response = await this.http.post('/api/user/checkout', data);
        // Clear both orders and subscriptions cache after checkout
        // to ensure fresh data when user returns from payment
        this.orders.clearCache();
        this.subscriptions.clearCache();
        return response.checkout;
    }
    /**
     * Get customer portal URL for managing subscriptions
     *
     * @example
     * const portal = await paycan.checkout.getPortalUrl();
     * window.location.href = portal.url;
     */
    async getPortalUrl(returnUrl) {
        const response = await this.http.post('/api/user/checkout/portal', returnUrl ? { return_url: returnUrl } : undefined);
        return response.portal;
    }
    /**
     * Preview checkout totals including tax
     *
     * Get a preview of the order total including tax calculations
     * before creating the actual checkout session.
     *
     * @example
     * const preview = await paycan.checkout.preview({
     *   product_price_id: 1,
     *   gateway: 'stripe',
     *   quantity: 2,
     *   billing_country: 'US',
     *   billing_state: 'CA'
     * });
     * console.log('Total:', preview.selected_price.final_price);
     */
    async preview(params) {
        const queryParams = new URLSearchParams();
        if (params.product_price_id) {
            queryParams.append('product_price_id', params.product_price_id.toString());
        }
        if (params.gateway) {
            queryParams.append('gateway', params.gateway);
        }
        if (params.quantity) {
            queryParams.append('quantity', params.quantity.toString());
        }
        if (params.billing_country) {
            queryParams.append('billing_country', params.billing_country);
        }
        if (params.billing_state) {
            queryParams.append('billing_state', params.billing_state);
        }
        return this.http.get(`/api/user/checkout/preview?${queryParams.toString()}`);
    }
}

/**
 * Products Resource
 *
 * Handle product-related operations
 */
class Products {
    constructor(http) {
        this.http = http;
    }
    /**
     * List all active products
     *
     * @param params - Query parameters for filtering, sorting, pagination, and includes
     * @param params.filter - Filter products (type)
     * @param params.include - Include related data (prices)
     * @param params.sort - Sort field (created_at, title) - prefix with - for descending
     * @param params.per_page - Items per page (1-100, default: 15)
     * @param params.page - Page number
     *
     * @example
     * // Basic usage
     * const products = await paycan.products.list();
     *
     * @example
     * // With filtering and sorting
     * const digitalProducts = await paycan.products.list({
     *   filter: { type: 'digital' },
     *   include: 'activePrices',
     *   sort: 'title',
     *   per_page: 20
     * });
     */
    async list(params) {
        return this.http.get('/api/user/products', params);
    }
    /**
     * Get a specific product by ID
     *
     * @param productId - The product ID
     * @param params - Query parameters
     * @param params.include - Include related data (prices)
     *
     * @example
     * const product = await paycan.products.get('product-123');
     *
     * @example
     * // With includes
     * const product = await paycan.products.get('product-123', {
     *   include: 'activePrices'
     * });
     */
    async get(productId, params) {
        return this.http.get(`/api/user/products/${productId}`, params);
    }
}

/**
 * Transactions Resource
 *
 * Handle transaction-related operations
 */
class Transactions {
    constructor(http) {
        this.http = http;
    }
    /**
     * List all transactions for the authenticated user
     *
     * @param params - Query parameters for filtering, sorting, pagination, and includes
     * @param params.filter - Filter transactions (type, status, gateway, created_after, created_before)
     * @param params.include - Include related data (order, subscription)
     * @param params.sort - Sort field (created_at, amount) - prefix with - for descending
     * @param params.per_page - Items per page (1-100, default: 15)
     * @param params.page - Page number
     *
     * @example
     * // Basic usage
     * const transactions = await paycan.transactions.list();
     *
     * @example
     * // With filtering and sorting
     * const completedTransactions = await paycan.transactions.list({
     *   filter: { status: 'completed', type: 'charge' },
     *   sort: '-created_at',
     *   per_page: 20
     * });
     */
    async list(params) {
        return this.http.get('/api/user/transactions', params);
    }
    /**
     * Get a specific transaction by ID
     *
     * @param transactionId - The transaction ID
     * @param params - Query parameters
     * @param params.include - Include related data (order, subscription)
     *
     * @example
     * const transaction = await paycan.transactions.get('txn-123');
     *
     * @example
     * // With includes
     * const transaction = await paycan.transactions.get('txn-123', {
     *   include: 'order'
     * });
     */
    async get(transactionId, params) {
        return this.http.get(`/api/user/transactions/${transactionId}`, params);
    }
}

/**
 * PayCan Checkout Modal Web Component
 *
 * A framework-agnostic checkout modal that can be embedded anywhere
 */
class CheckoutModal {
    constructor(sdk, options) {
        this.overlay = null;
        this.modal = null;
        this.previewData = null;
        this.selectedPriceId = null;
        this.selectedGateway = 'stripe';
        this.loading = false;
        /**
         * Handle escape key press
         */
        this.handleEscapeKey = (e) => {
            if (e.key === 'Escape') {
                this.close();
            }
        };
        this.sdk = sdk;
        this.options = options;
        this.selectedPriceId = options.priceId || null;
    }
    /**
     * Open the modal and load preview data
     */
    async open() {
        try {
            // Load preview data
            await this.loadPreview();
            // Create and show modal
            this.createModal();
            this.show();
        }
        catch (error) {
            this.handleError(error);
        }
    }
    /**
     * Close and destroy the modal
     */
    close() {
        if (this.overlay) {
            this.overlay.remove();
            this.overlay = null;
            this.modal = null;
        }
        if (this.options.onCancel) {
            this.options.onCancel();
        }
    }
    /**
     * Load preview data from API
     */
    async loadPreview() {
        const params = {
            quantity: 1,
        };
        if (this.options.productId) {
            params.product_id = this.options.productId;
            if (this.selectedPriceId) {
                params.selected_price_id = this.selectedPriceId;
            }
        }
        else if (this.selectedPriceId) {
            params.product_price_id = this.selectedPriceId;
        }
        this.previewData = await this.sdk.checkout.preview(params);
        // Set default selected price if not set
        if (!this.selectedPriceId && this.previewData.selected_price) {
            this.selectedPriceId = this.previewData.selected_price.id;
        }
        // Set default gateway from available payment methods
        if (this.previewData.payment_methods.length > 0) {
            this.selectedGateway = this.previewData.payment_methods[0].key;
        }
    }
    /**
     * Create modal DOM structure
     */
    createModal() {
        const theme = this.options.theme || 'light';
        // Create overlay
        this.overlay = document.createElement('div');
        this.overlay.className = `paycan-modal-overlay paycan-theme-${theme}`;
        this.overlay.style.cssText = `
      position: fixed;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      background: rgba(0, 0, 0, 0.5);
      display: flex;
      align-items: center;
      justify-content: center;
      z-index: 999999;
      padding: 1rem;
      opacity: 0;
      transition: opacity 0.2s ease-in-out;
    `;
        // Create modal container
        this.modal = document.createElement('div');
        this.modal.className = 'paycan-modal';
        this.modal.style.cssText = `
      background: ${theme === 'dark' ? '#1f2937' : '#ffffff'};
      border-radius: 12px;
      box-shadow: 0 20px 25px -5px rgba(0, 0, 0, 0.1), 0 10px 10px -5px rgba(0, 0, 0, 0.04);
      max-width: 500px;
      width: 100%;
      max-height: 90vh;
      overflow-y: auto;
      color: ${theme === 'dark' ? '#f9fafb' : '#111827'};
      transform: scale(0.95);
      transition: transform 0.2s ease-in-out;
    `;
        this.modal.innerHTML = this.getModalContent();
        this.overlay.appendChild(this.modal);
        document.body.appendChild(this.overlay);
        // Add event listeners
        this.attachEventListeners();
        // Close on overlay click
        this.overlay.addEventListener('click', (e) => {
            if (e.target === this.overlay) {
                this.close();
            }
        });
        // Close on Escape key
        document.addEventListener('keydown', this.handleEscapeKey);
    }
    /**
     * Get modal HTML content
     */
    getModalContent() {
        if (!this.previewData) {
            return '<div style="padding: 2rem; text-align: center;">Loading...</div>';
        }
        const theme = this.options.theme || 'light';
        const isDark = theme === 'dark';
        const isAuthenticated = this.sdk.isAuthenticated();
        const product = this.previewData.product;
        const selectedPrice = this.previewData.selected_price;
        const prices = this.previewData.prices;
        const paymentMethods = this.previewData.payment_methods;
        return `
      <div style="padding: 1.5rem;">
        <!-- Header -->
        <div style="display: flex; align-items: center; justify-content: space-between; margin-bottom: 1.5rem;">
          <h2 style="font-size: 1.5rem; font-weight: 700; margin: 0; color: ${isDark ? '#f9fafb' : '#111827'};">
            🛒 Checkout
          </h2>
          <button
            class="paycan-close-btn"
            style="background: none; border: none; font-size: 1.5rem; cursor: pointer; padding: 0; color: ${isDark ? '#9ca3af' : '#6b7280'}; line-height: 1;"
            aria-label="Close"
          >
            ×
          </button>
        </div>

        <!-- Product Info -->
        <div style="background: ${isDark ? '#374151' : '#f9fafb'}; border: 1px solid ${isDark ? '#4b5563' : '#e5e7eb'}; border-radius: 8px; padding: 1rem; margin-bottom: 1.5rem;">
          <h3 style="font-size: 1.125rem; font-weight: 600; margin: 0 0 0.5rem 0; color: ${isDark ? '#f9fafb' : '#111827'};">
            ${this.escapeHtml(product.name)}
          </h3>
          ${product.description ? `<p style="font-size: 0.875rem; color: ${isDark ? '#d1d5db' : '#6b7280'}; margin: 0;">${this.escapeHtml(product.description)}</p>` : ''}
          <div style="margin-top: 0.75rem;">
            <span style="display: inline-block; background: ${this.getProductTypeBg(product.type, isDark)}; color: ${this.getProductTypeColor(product.type, isDark)}; padding: 0.25rem 0.75rem; border-radius: 9999px; font-size: 0.75rem; font-weight: 500;">
              ${this.escapeHtml(product.type)}
            </span>
          </div>
        </div>

        ${!isAuthenticated ? this.getGuestEmailField(isDark) : ''}

        ${prices.length > 1 ? this.getPriceSelectionField(prices, isDark) : ''}

        ${selectedPrice ? this.getPriceBreakdown(selectedPrice, isDark) : ''}

        ${paymentMethods.length > 0 ? this.getPaymentMethodsField(paymentMethods, isDark) : ''}

        <!-- Security Notice -->
        <div style="background: ${isDark ? '#1f2937' : '#f9fafb'}; border: 1px solid ${isDark ? '#374151' : '#e5e7eb'}; border-radius: 6px; padding: 0.75rem; margin-bottom: 1.5rem;">
          <p style="font-size: 0.75rem; color: ${isDark ? '#9ca3af' : '#6b7280'}; margin: 0;">
            🔒 Your payment information is secure and encrypted. We never store your card details.
          </p>
        </div>

        <!-- Actions -->
        <div style="display: flex; gap: 0.75rem; justify-content: flex-end;">
          <button
            class="paycan-cancel-btn"
            style="background: ${isDark ? '#374151' : '#ffffff'}; color: ${isDark ? '#f9fafb' : '#111827'}; border: 1px solid ${isDark ? '#4b5563' : '#d1d5db'}; padding: 0.625rem 1.25rem; border-radius: 6px; font-size: 0.875rem; font-weight: 500; cursor: pointer; transition: all 0.15s;"
          >
            Cancel
          </button>
          <button
            class="paycan-checkout-btn"
            style="background: #3b82f6; color: white; border: none; padding: 0.625rem 1.5rem; border-radius: 6px; font-size: 0.875rem; font-weight: 500; cursor: pointer; transition: all 0.15s;"
            ${this.loading ? 'disabled' : ''}
          >
            ${this.loading ? 'Processing...' : selectedPrice ? `Pay ${this.formatPrice(selectedPrice.final_price, selectedPrice.currency)}` : 'Continue'}
          </button>
        </div>
      </div>
    `;
    }
    /**
     * Get guest email field HTML
     */
    getGuestEmailField(isDark) {
        return `
      <div style="margin-bottom: 1.5rem;">
        <label style="display: block; font-size: 0.875rem; font-weight: 500; margin-bottom: 0.5rem; color: ${isDark ? '#f9fafb' : '#111827'};">
          Email Address <span style="color: #ef4444;">*</span>
        </label>
        <input
          type="email"
          class="paycan-email-input"
          placeholder="your@email.com"
          required
          style="width: 100%; padding: 0.625rem; border: 1px solid ${isDark ? '#4b5563' : '#d1d5db'}; border-radius: 6px; font-size: 0.875rem; background: ${isDark ? '#374151' : '#ffffff'}; color: ${isDark ? '#f9fafb' : '#111827'}; box-sizing: border-box;"
        />
        <p style="font-size: 0.75rem; color: ${isDark ? '#9ca3af' : '#6b7280'}; margin: 0.375rem 0 0 0;">
          We'll send your receipt and purchase details to this email.
        </p>
      </div>
    `;
    }
    /**
     * Get price selection field HTML
     */
    getPriceSelectionField(prices, isDark) {
        if (prices.length < 3) {
            // Radio buttons
            return `
        <div style="margin-bottom: 1.5rem;">
          <label style="display: block; font-size: 0.875rem; font-weight: 500; margin-bottom: 0.75rem; color: ${isDark ? '#f9fafb' : '#111827'};">
            Select Plan
          </label>
          <div style="display: flex; flex-direction: column; gap: 0.75rem;">
            ${prices.map(price => `
              <label style="display: flex; align-items: center; padding: 1rem; border: 2px solid ${this.selectedPriceId === price.id ? '#3b82f6' : (isDark ? '#4b5563' : '#e5e7eb')}; background: ${this.selectedPriceId === price.id ? (isDark ? '#1e3a8a' : '#eff6ff') : (isDark ? '#374151' : '#ffffff')}; border-radius: 8px; cursor: pointer; transition: all 0.15s;">
                <input
                  type="radio"
                  name="price"
                  value="${price.id}"
                  class="paycan-price-radio"
                  ${this.selectedPriceId === price.id ? 'checked' : ''}
                  style="margin-right: 0.75rem;"
                />
                <div style="flex: 1;">
                  <div style="display: flex; justify-content: space-between; align-items: center;">
                    <span style="font-weight: 600; color: ${isDark ? '#f9fafb' : '#111827'};">${this.escapeHtml(price.name)}</span>
                    <span style="font-weight: 700; font-size: 1.125rem; color: ${isDark ? '#f9fafb' : '#111827'};">${this.formatPrice(price.final_price, price.currency)}</span>
                  </div>
                  <div style="font-size: 0.75rem; color: ${isDark ? '#9ca3af' : '#6b7280'}; margin-top: 0.25rem;">
                    ${this.getBillingPeriodText(price.billing_period)}${price.trial_days ? ` • ${price.trial_days} days free trial` : ''}
                  </div>
                </div>
              </label>
            `).join('')}
          </div>
        </div>
      `;
        }
        else {
            // Dropdown
            return `
        <div style="margin-bottom: 1.5rem;">
          <label style="display: block; font-size: 0.875rem; font-weight: 500; margin-bottom: 0.5rem; color: ${isDark ? '#f9fafb' : '#111827'};">
            Select Plan
          </label>
          <select
            class="paycan-price-select"
            style="width: 100%; padding: 0.625rem; border: 1px solid ${isDark ? '#4b5563' : '#d1d5db'}; border-radius: 6px; font-size: 0.875rem; background: ${isDark ? '#374151' : '#ffffff'}; color: ${isDark ? '#f9fafb' : '#111827'}; cursor: pointer;"
          >
            ${prices.map(price => `
              <option value="${price.id}" ${this.selectedPriceId === price.id ? 'selected' : ''}>
                ${this.escapeHtml(price.name)} - ${this.formatPrice(price.final_price, price.currency)} ${this.getBillingPeriodText(price.billing_period)}
              </option>
            `).join('')}
          </select>
        </div>
      `;
        }
    }
    /**
     * Get price breakdown HTML
     */
    getPriceBreakdown(price, isDark) {
        return `
      <div style="background: ${isDark ? '#111827' : '#f9fafb'}; border: 1px solid ${isDark ? '#374151' : '#e5e7eb'}; border-radius: 8px; padding: 1rem; margin-bottom: 1.5rem;">
        <div style="display: flex; justify-content: space-between; margin-bottom: 0.5rem;">
          <span style="font-size: 0.875rem; color: ${isDark ? '#9ca3af' : '#6b7280'};">Subtotal</span>
          <span style="font-size: 0.875rem; font-weight: 500; color: ${isDark ? '#f9fafb' : '#111827'};">${this.formatPrice(price.subtotal, price.currency)}</span>
        </div>
        ${price.tax_amount > 0 ? `
          <div style="display: flex; justify-content: space-between; margin-bottom: 0.5rem;">
            <span style="font-size: 0.875rem; color: ${isDark ? '#9ca3af' : '#6b7280'};">Tax (${(this.previewData.tax.rate * 100).toFixed(1)}%)</span>
            <span style="font-size: 0.875rem; font-weight: 500; color: ${isDark ? '#f9fafb' : '#111827'};">${this.formatPrice(price.tax_amount, price.currency)}</span>
          </div>
        ` : ''}
        <div style="border-top: 1px solid ${isDark ? '#374151' : '#e5e7eb'}; margin: 0.75rem 0; padding-top: 0.75rem;"></div>
        <div style="display: flex; justify-content: space-between; align-items: center;">
          <span style="font-weight: 600; color: ${isDark ? '#f9fafb' : '#111827'};">Total</span>
          <div style="text-align: right;">
            <div style="font-size: 1.5rem; font-weight: 700; color: ${isDark ? '#f9fafb' : '#111827'};">${this.formatPrice(price.final_price, price.currency)}</div>
            <div style="font-size: 0.75rem; color: ${isDark ? '#9ca3af' : '#6b7280'};">${this.getBillingPeriodText(price.billing_period)}</div>
          </div>
        </div>
      </div>
    `;
    }
    /**
     * Get payment methods field HTML
     */
    getPaymentMethodsField(methods, isDark) {
        return `
      <div style="margin-bottom: 1.5rem;">
        <label style="display: block; font-size: 0.875rem; font-weight: 500; margin-bottom: 0.75rem; color: ${isDark ? '#f9fafb' : '#111827'};">
          Payment Method
        </label>
        <div style="display: flex; flex-direction: column; gap: 0.75rem;">
          ${methods.map(method => `
            <label style="display: flex; align-items: center; padding: 1rem; border: 2px solid ${this.selectedGateway === method.key ? '#3b82f6' : (isDark ? '#4b5563' : '#e5e7eb')}; background: ${this.selectedGateway === method.key ? (isDark ? '#1e3a8a' : '#eff6ff') : (isDark ? '#374151' : '#ffffff')}; border-radius: 8px; cursor: pointer; transition: all 0.15s;">
              <input
                type="radio"
                name="gateway"
                value="${method.key}"
                class="paycan-gateway-radio"
                ${this.selectedGateway === method.key ? 'checked' : ''}
                style="margin-right: 0.75rem;"
              />
              <div style="flex: 1;">
                <div style="font-weight: 600; color: ${isDark ? '#f9fafb' : '#111827'}; margin-bottom: 0.25rem;">${this.escapeHtml(method.name)}</div>
                ${method.description ? `<div style="font-size: 0.75rem; color: ${isDark ? '#9ca3af' : '#6b7280'};">${this.escapeHtml(method.description)}</div>` : ''}
              </div>
              ${method.supports_subscriptions ? '<span style="font-size: 0.625rem; background: #10b981; color: white; padding: 0.125rem 0.5rem; border-radius: 9999px; font-weight: 500;">Subscriptions</span>' : ''}
            </label>
          `).join('')}
        </div>
      </div>
    `;
    }
    /**
     * Attach event listeners to modal elements
     */
    attachEventListeners() {
        if (!this.modal)
            return;
        // Close button
        const closeBtn = this.modal.querySelector('.paycan-close-btn');
        closeBtn?.addEventListener('click', () => this.close());
        // Cancel button
        const cancelBtn = this.modal.querySelector('.paycan-cancel-btn');
        cancelBtn?.addEventListener('click', () => this.close());
        // Checkout button
        const checkoutBtn = this.modal.querySelector('.paycan-checkout-btn');
        checkoutBtn?.addEventListener('click', () => this.handleCheckout());
        // Price radio buttons
        const priceRadios = this.modal.querySelectorAll('.paycan-price-radio');
        priceRadios.forEach(radio => {
            radio.addEventListener('change', (e) => {
                const target = e.target;
                this.handlePriceChange(parseInt(target.value));
            });
        });
        // Price dropdown
        const priceSelect = this.modal.querySelector('.paycan-price-select');
        priceSelect?.addEventListener('change', (e) => {
            const target = e.target;
            this.handlePriceChange(parseInt(target.value));
        });
        // Gateway radio buttons
        const gatewayRadios = this.modal.querySelectorAll('.paycan-gateway-radio');
        gatewayRadios.forEach(radio => {
            radio.addEventListener('change', (e) => {
                const target = e.target;
                this.selectedGateway = target.value;
            });
        });
    }
    /**
     * Handle price change
     */
    async handlePriceChange(priceId) {
        this.selectedPriceId = priceId;
        try {
            // Reload preview with new price
            await this.loadPreview();
            // Re-render modal
            if (this.modal) {
                this.modal.innerHTML = this.getModalContent();
                this.attachEventListeners();
            }
        }
        catch (error) {
            this.handleError(error);
        }
    }
    /**
     * Handle checkout button click
     */
    async handleCheckout() {
        if (this.loading)
            return;
        try {
            this.loading = true;
            this.updateCheckoutButton();
            // Validate email if guest
            let billingEmail;
            if (!this.sdk.isAuthenticated()) {
                const emailInput = this.modal?.querySelector('.paycan-email-input');
                if (!emailInput || !emailInput.value || !this.isValidEmail(emailInput.value)) {
                    throw new Error('Please enter a valid email address');
                }
                billingEmail = emailInput.value;
            }
            // Create checkout session
            const checkoutData = {
                product_price_id: this.selectedPriceId,
                gateway: this.selectedGateway,
                quantity: 1,
            };
            if (this.previewData) {
                checkoutData.product_id = this.previewData.product.id;
            }
            if (billingEmail) {
                checkoutData.billing_email = billingEmail;
            }
            const session = await this.sdk.checkout.create(checkoutData);
            // Call success callback or redirect
            if (this.options.onSuccess) {
                this.options.onSuccess(session.checkout_url);
            }
            else {
                window.location.href = session.checkout_url;
            }
            this.close();
        }
        catch (error) {
            this.loading = false;
            this.updateCheckoutButton();
            this.handleError(error);
        }
    }
    /**
     * Update checkout button state
     */
    updateCheckoutButton() {
        const btn = this.modal?.querySelector('.paycan-checkout-btn');
        if (btn) {
            btn.disabled = this.loading;
            btn.textContent = this.loading ? 'Processing...' :
                this.previewData?.selected_price ?
                    `Pay ${this.formatPrice(this.previewData.selected_price.final_price, this.previewData.selected_price.currency)}` :
                    'Continue';
        }
    }
    /**
     * Show modal with animation
     */
    show() {
        if (this.overlay && this.modal) {
            // Trigger reflow for animation
            this.overlay.offsetHeight;
            this.overlay.style.opacity = '1';
            this.modal.style.transform = 'scale(1)';
        }
    }
    /**
     * Handle errors
     */
    handleError(error) {
        console.error('[PayCan Checkout Modal]', error);
        if (this.options.onError) {
            this.options.onError(error);
        }
        else {
            alert(`Checkout Error: ${error.message}`);
        }
    }
    /**
     * Format price
     */
    formatPrice(amount, currency) {
        return new Intl.NumberFormat('en-US', {
            style: 'currency',
            currency: currency || 'USD',
        }).format(amount);
    }
    /**
     * Get billing period text
     */
    getBillingPeriodText(period) {
        if (period === 'once')
            return 'one-time payment';
        return `per ${period}`;
    }
    /**
     * Get product type background color
     */
    getProductTypeBg(type, isDark) {
        const colors = {
            digital: isDark ? '#1e3a8a' : '#dbeafe',
            physical: isDark ? '#065f46' : '#d1fae5',
            service: isDark ? '#581c87' : '#f3e8ff',
            subscription: isDark ? '#9a3412' : '#fed7aa',
        };
        return colors[type] || (isDark ? '#374151' : '#f3f4f6');
    }
    /**
     * Get product type text color
     */
    getProductTypeColor(type, isDark) {
        const colors = {
            digital: isDark ? '#93c5fd' : '#1e40af',
            physical: isDark ? '#6ee7b7' : '#047857',
            service: isDark ? '#d8b4fe' : '#6b21a8',
            subscription: isDark ? '#fdba74' : '#c2410c',
        };
        return colors[type] || (isDark ? '#d1d5db' : '#4b5563');
    }
    /**
     * Validate email
     */
    isValidEmail(email) {
        return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);
    }
    /**
     * Escape HTML to prevent XSS
     */
    escapeHtml(text) {
        const div = document.createElement('div');
        div.textContent = text;
        return div.innerHTML;
    }
    /**
     * Cleanup
     */
    destroy() {
        document.removeEventListener('keydown', this.handleEscapeKey);
        this.close();
    }
}

/**
 * PayCan SDK - Main Client
 *
 * Official JavaScript SDK for PayCan payment integration
 */
class PayCan {
    /**
     * Create a new PayCan instance
     *
     * @example
     * const paycan = new PayCan({
     *   apiUrl: 'https://pay.yourapp.com',
     *   debug: true
     * });
     */
    constructor(config) {
        this.validateConfig(config);
        this.http = new HttpClient(config);
        this.orders = new Orders(this.http, config);
        this.subscriptions = new Subscriptions(this.http, config);
        this.products = new Products(this.http);
        this.transactions = new Transactions(this.http);
        this.checkout = new Checkout(this.http, this.orders, this.subscriptions);
    }
    /**
     * Get current user information
     *
     * @example
     * const { user } = await paycan.me();
     */
    async me() {
        return this.http.get('/api/user/me');
    }
    /**
     * Set the PayCan user token directly
     *
     * This is the primary method for authentication. Your app should:
     * 1. Authenticate the user in YOUR system
     * 2. Call your backend to get a PayCan token for this user
     * 3. Pass the token to this method
     *
     * @example
     * // Get token from your backend however you want
     * const response = await fetch('/api/get-paycan-token');
     * const { token } = await response.json();
     *
     * // Set the token
     * paycan.setUserToken(token);
     *
     * // Now you can make authenticated requests
     * const orders = await paycan.orders.list();
     */
    setUserToken(token) {
        if (!token) {
            throw new Error('Token is required');
        }
        this.http.setToken(token);
        this.log('User token set successfully');
    }
    /**
     * Helper method to authenticate and get token from your backend
     *
     * This is optional - you can use setUserToken() instead if you prefer
     * to handle authentication yourself.
     *
     * @example
     * // Option 1: Bearer token (JWT in Authorization header)
     * await paycan.authenticate({
     *   endpoint: '/api/paycan/token',
     *   type: 'bearer',
     *   headers: { 'Authorization': `Bearer ${yourJWT}` }
     * });
     *
     * @example
     * // Option 2: Cookie-based (sends cookies automatically)
     * await paycan.authenticate({
     *   endpoint: '/api/paycan/token',
     *   type: 'cookie'
     * });
     *
     * @example
     * // Option 3: Custom authentication
     * await paycan.authenticate({
     *   endpoint: '/api/paycan/token',
     *   type: 'custom',
     *   headers: { 'X-Custom-Auth': 'your-value' }
     * });
     */
    async authenticate(config) {
        try {
            const headers = {
                'Content-Type': 'application/json',
                ...config.headers,
            };
            const fetchOptions = {
                method: 'POST',
                headers,
            };
            // Add credentials for cookie-based auth
            if (config.type === 'cookie') {
                fetchOptions.credentials = 'include';
            }
            // Add body data if provided
            if (config.data) {
                fetchOptions.body = JSON.stringify(config.data);
            }
            const response = await fetch(config.endpoint, fetchOptions);
            if (!response.ok) {
                throw new Error(`Authentication failed: ${response.statusText}`);
            }
            const data = await response.json();
            if (!data.token) {
                throw new Error('No token received from auth endpoint');
            }
            this.http.setToken(data.token);
            this.log('User authenticated successfully');
            return data;
        }
        catch (error) {
            this.log('Authentication error:', error);
            throw error;
        }
    }
    /**
     * Check if user is authenticated
     *
     * @example
     * if (paycan.isAuthenticated()) {
     *   // User is logged in
     * }
     */
    isAuthenticated() {
        return this.http.getToken() !== null;
    }
    /**
     * Get current token (useful for debugging)
     */
    getToken() {
        return this.http.getToken();
    }
    /**
     * Open checkout modal for a product
     *
     * This will display a modal with all available prices for the product,
     * allowing the user to select a price and payment gateway before checkout.
     *
     * @example
     * // Basic usage
     * paycan.openCheckoutModal(123);
     *
     * @example
     * // With options
     * paycan.openCheckoutModal(123, {
     *   theme: 'dark',
     *   onSuccess: (checkoutUrl) => {
     *     console.log('Redirecting to:', checkoutUrl);
     *     window.location.href = checkoutUrl;
     *   },
     *   onCancel: () => {
     *     console.log('User cancelled checkout');
     *   },
     *   onError: (error) => {
     *     console.error('Checkout error:', error);
     *   }
     * });
     */
    openCheckoutModal(productId, options = {}) {
        const modalOptions = {
            productId,
            theme: options.theme || 'light',
            onSuccess: options.onSuccess,
            onCancel: options.onCancel,
            onError: options.onError,
        };
        const modal = new CheckoutModal(this, modalOptions);
        modal.open();
        this.log('Opening checkout modal for product:', productId);
        return modal;
    }
    /**
     * Open checkout modal for a specific price
     *
     * This will display a modal pre-configured for a specific price,
     * allowing the user to select a payment gateway before checkout.
     *
     * @example
     * // Basic usage
     * paycan.openCheckoutModalPrice(456);
     *
     * @example
     * // With options and callbacks
     * paycan.openCheckoutModalPrice(456, {
     *   theme: 'light',
     *   onSuccess: (checkoutUrl) => {
     *     // Custom success handler
     *     window.location.href = checkoutUrl;
     *   },
     *   onCancel: () => {
     *     console.log('Checkout cancelled');
     *   }
     * });
     */
    openCheckoutModalPrice(priceId, options = {}) {
        const modalOptions = {
            priceId,
            theme: options.theme || 'light',
            onSuccess: options.onSuccess,
            onCancel: options.onCancel,
            onError: options.onError,
        };
        const modal = new CheckoutModal(this, modalOptions);
        modal.open();
        this.log('Opening checkout modal for price:', priceId);
        return modal;
    }
    /**
     * Validate configuration
     */
    validateConfig(config) {
        if (!config.apiUrl) {
            throw new Error('PayCan: apiUrl is required');
        }
        // Ensure URL doesn't have trailing slash
        config.apiUrl = config.apiUrl.replace(/\/$/, '');
    }
    /**
     * Debug logging
     */
    log(...args) {
        const config = this.http['config'];
        if (config.debug) {
            console.log('[PayCan SDK]', ...args);
        }
    }
}

export { CheckoutModal, PayCan, PayCan as default };
